x-pack/plugins/files/server/routes/file_kind/enhance_router.ts

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { RequestHandler, RouteMethod, RouteRegistrar } from '@kbn/core/server';
+
+import { FilesRouter } from '../types';
+import { FileKindRouter, FileKindsRequestHandlerContext } from './types';
+
+interface Args {
+  router: FilesRouter;
+  fileKind: string;
+}
+
+type FileKindHandler = RequestHandler<unknown, unknown, unknown, FileKindsRequestHandlerContext>;
+
+/**
+ * Wraps {@link FilesRouter}, adding a middle man for injecting file-kind into
+ * route handler context
+ */
+export function enhanceRouter({ router, fileKind }: Args): FileKindRouter {
```

Comment:
Q: What does this `enhanceRouter` do? Can it be achieved without reflection?

Reply:
It wraps the router we get from core so that we can intercept any handlers and make sure that `fileKind` is added to the context passed to the handler.

The same can be done by (very simplified code):

```ts

const newRouter = {
  ...originalRouter,
  get: (handler) => originalRouter.get(wrapHandler(handler))
  post: (handler) => originalRouter.post(wrapHandler(handler))
  put: (handler) => originalRouter.put(wrapHandler(handler))
  ...
}
```

Or we can make the handlers handler factories by making them: `const handler = (fileKind: string) => (ctx, req, res) => ...`.

I opted for this approach because it will work if the `originalRouter` is a class or an object and it helps us keep the handler code cleaner - since there is more handler code than this kind of code it seemed worth it.